Data Request Pages (2nd Attempt)

Allcharles opened the pull request:
# Second implementation of the Data Request pages

Simplify the changes proposed by #968 

## Changes

- Update menu with the following changes:
  - Enabled the use of multiple widgets on the menu
  - Created `MenuModal` object which creates an action on the menu which can open a modal containing a component
- Implemented Data Request Page
- Created Download Annotations modal and added link in site and point pages
- Created optional show resolvers, this allows us to request a model without throwing an error if it does not exist. Here is the difference in code between show and showOptional:

```typescript
PointAnnotationsComponent.linkComponentToPageInfo({
  category: pointsCategory,
  resolvers: {
    [projectKey]: projectResolvers.show,
    [regionKey]: regionResolvers.show,
    [siteKey]: siteResolvers.show,
  },
}).andMenuRoute(pointAnnotationsMenuItem);

PointAnnotationsComponent.linkComponentToPageInfo({
  category: pointsCategory,
  resolvers: {
    [projectKey]: projectResolvers.showOptional,
    [regionKey]: regionResolvers.showOptional,
    [siteKey]: siteResolvers.showOptional,
  },
}).andMenuRoute(pointAnnotationsMenuItem);
```

## Problems

- Modals not tested
- Menu not tested
- Data Request service not tested
- Did not implement user download annotations page (@atruskie should I include that in this PR?)
  - Currently the link in the users account only points to the data request page telling them how to download annotations for a specific site
- Annotation download form fails because authToken is not sent in the request

## Issues

- Closes #614 

## Visual Changes

### Data Request Page

![image](https://user-images.githubusercontent.com/3955116/117411392-db866b80-af56-11eb-9ff1-e0d75de6f695.png)

### Site Download Annotations Page

![image](https://user-images.githubusercontent.com/3955116/120599868-66418400-c48b-11eb-9b1f-35a8a4b08bd6.png)

### Point Download Annotations Page

![image](https://user-images.githubusercontent.com/3955116/120599899-7194af80-c48b-11eb-90dd-55dde05b2636.png)

## Final Checklist

- [x] Assign reviewers if you have permission
- [x] Assign labels if you have permission
- [x] Link issues related to PR
- [x] Ensure project linter is not producing any warnings (`npm run lint`)
- [x] Ensure build is passing on all browsers (`npm run test:all`)
- [ ] Ensure CI build is passing
- [ ] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

Allcharles committed:
Implemented Data Request pages

- Created optional show resolver
- Implemented data request page
- General template for modal design
- Working annotation download modal
- Update menus to accept multiple widgets
- Merge modals into menu links list

atruskie commented on src/app/components/data-request/data-request.component.ts:
```diff
@@ -20,60 +24,61 @@ import { fields as annotationFields } from "./download-annotations.schema.json";
+          related to the audio recordings on this website.
+        </p>
+        <p>
+          You <strong>do not need</strong> to use this form if you need the
+          standard <strong>annotations CSV</strong> download.
```
Add instructions for standard download:

> 1. Open the site or point page you're interested in
> 2. Use the _Download Annotations_ button to download annotations

atruskie commented on src/app/components/sites/pages/annotations/point.component.ts:
```diff
@@ -0,0 +1,60 @@
+import { Component, OnInit } from "@angular/core";
```
this file probably should not be named `point` ... _point_ is a user term, not a technical term

atruskie reviewed (changes requested):
Okay, the reason I suggested a modal component for the annotation downloader is so that you could embed it on the region/site pages and bind to the existing data models without needing to load them on another page. Adding more pages doesn't improve on the old design...

Otherwise, a fine set of changes. 

I think it's worth removing the redundant annotation download pages though, it can simplify the design.

For the pages added, they're almost duplicates, which is not ideal either.

atruskie commented on src/app/services/baw-api/site/sites.service.ts:
```diff
@@ -90,6 +92,19 @@ export class SitesService extends StandardApi<Site, [IdOr<Project>]> {
+  /**
+   * Retrieve path to generated a CSV file containing all of the annotations for a site.
+   * Insert into the `[href]` of an anchor HTML element.
+   * TODO Append authToken so that requests are authenticated
```
intentionally not done?

atruskie reviewed (approved):
great changes

Allcharles committed:
Implemented tests for Data Request Pages (#1129)

- Implemented site service tests
- Implemented annotation-download component tests
- Implemented menu component tests
- Implemented secondary/action tests
- Changed type of widgets input from array to List to match other inputs
